README.md
# gfk-period 

A class that provides a DTO to represent a time period
and methods to render the dates of which it consists

It will always expand the provided dates for start and end to be the start and end of the
period they are in.
Start date will always be monday, the first of the month, the beginning of the quarter or the year
End date will always be sunday, the last of the month, the end of the quarter or the year

Only if PeriodMode DAYS is chosen, the dates will keep their exact values

## Install

### As NPM package
```bash
$ npm install --save gfke/period
```
then
```javascript
 require('gfk-period');
```

### As bower component
```bash
$ bower install --save gfk-period
```
then 
```html
<script src="/lib/period.js"></script>
<!--You will need to include moment and moment-range libraries yourself-->
```
or 
```html
<script src="/dist/period.js"></script>
<!--Already includes moment and moment-range libraries as well-->
```


## Usage

```javascript
var period = require('gfk-period');
var periodInstance = period.createPeriod(periodMode, start, end);
```

## API

```javascript
createPeriod (periodMode, start, end)
//Factory method to create a period instance
```

```javascript
PERIOD_MODES = {
    DAYS: 'd',
    WEEKS: 'w',
    MONTHS: 'm',
    QUARTERS: 'q',
    HALF_YEAR: 'h',
    YEARS: 'y'
  };
  //enum for all available PeriodModes
```

```javascript
Period.prototype.getLongPeriodLabel();
//Get a label representing the complete period
```

```javascript
Period.prototype.isEqual
```

```javascript
Period.prototype.getLongStringForStart
```

```javascript
Period.prototype.getLongStringForEnd
```

```javascript
Period.prototype.getLongStringFormat
```

```javascript
Period.prototype.getShortStringFormat
```

```javascript
Period.prototype.getGroupStringFormat
```

```javascript
Period.prototype.getValueAsObjects
```

## License

Copyright (c) 2015. Not licensed under the any license.


CHANGELOG.md
# 1.4.1 (2015-06-05)

- Fixed utc conversion problem

# 1.4.0 (2015-05-06)

- Group weeks by week of year

# 1.3.0 (2015-04-29)

- Add Year to GroupLabel of PERIOD_MODE.DAYS

# 1.2.0 (2015-04-14)

- Expose getMomentRangeIterator

# 1.1.0 (2015-04-14)

- Add checksum to output objects

# 1.0.0 (2015-04-14)

- Change checksum calculation

# 0.2.2 (2015-02-17)

- Fix sunday/monday handling

# 0.2.1 (2015-02-17)

- Fix Quarter group label

# 0.2.0 (2015-02-11)

- Implement expanding time ranges to at least one period

# 0.1.0 (2015-02-11)

- Implement isDirty method to identify changes on the period

# 0.0.4 (2015-02-11)

- Remove HALF_YEAR period mode

# 0.0.1 ... 0.0.3 (2015-01-27)

- Initial versions


test/period_test.js
'use strict';

var period = require('../lib/period');
var moment = require('moment');
var periodModes = period.PERIOD_MODES;


function indicesOf(array, element) {
    var idx = array.indexOf(element),
        indices = [];
    while (idx != -1) {
        indices.push(idx);
        idx = array.indexOf(element, idx + 1);
    }
    return indices
}

function countDatesInMoments(array, value) {
    var count = 0;
    for (var i = 0; i < array.length; i++) {
        var obj = array[i];
        if (obj.date() === value) count++;
    }
    return count;
}

function countMonthsInMoments(array, value) {
    var count = 0;
    for (var i = 0; i < array.length; i++) {
        var obj = array[i];
        if (obj.month() === value) count++;
    }
    return count;
}

describe('Test functions', function () {
    it('countDatesInMoments should return correct number of dates', function () {
        var testArray = [
            moment('2015-01-01'),
            moment('2015-02-02'),
            moment('2015-03-02')
        ];
        var countFirstOfMonths = countDatesInMoments(testArray, 1);
        var countSecondOfMonths = countDatesInMoments(testArray, 2);
        var countThirdOfMonths = countDatesInMoments(testArray, 3);

        expect(countFirstOfMonths).toEqual(1);
        expect(countSecondOfMonths).toEqual(2);
        expect(countThirdOfMonths).toEqual(0);
    });
});

describe('Helper functions', function () {
    it('getNumberOfDaysInMonth should return number of days in provided month', function () {
        var daysInJanuary2015 = 31;
        expect(period._getNumberOfDaysInMonth(2015, 0)).toEqual(daysInJanuary2015);
    });

    it('setPeriodOnMoment should mutate provided moments days', function () {
        var momentToUse = moment('2015-01-01');
        var expectedDate = 5;
        period._setPeriodOnMoment(periodModes.DAYS, momentToUse, expectedDate);
        expect(momentToUse.date()).toEqual(expectedDate);
    });

    it('setPeriodOnMoment should mutate provided moments month', function () {
        var momentToUse = moment('2015-01-01');
        var expectedMonth = 11;
        period._setPeriodOnMoment(periodModes.MONTHS, momentToUse, expectedMonth);
        expect(momentToUse.month()).toEqual(expectedMonth);
    });

    it('setPeriodOnMoment should mutate provided moments years', function () {
        var momentToUse = moment('2015-01-01');
        var expectedYear = 11;
        period._setPeriodOnMoment(periodModes.YEARS, momentToUse, expectedYear);
        expect(momentToUse.year()).toEqual(expectedYear);
    });

    it('setPeriodOnMoment should mutate provided moments quarters', function () {
        var momentToUse = moment('2015-01-01');
        var expectedQuarter = 3;
        var expectedMonth = 6;
        period._setPeriodOnMoment(periodModes.QUARTERS, momentToUse, expectedQuarter);

        expect(momentToUse.quarter()).toEqual(expectedQuarter);
        expect(momentToUse.month()).toEqual(expectedMonth);
    });

    describe('isNewPeriodGroup', function () {
        it('should return true for first of month if periodModes.DAYS', function () {
            var momentToUse = moment('2015-01-01'),
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.DAYS, momentToUse)
            expect(isNewGroup).toBeTruthy();
        });

        it('should return false for second of month if periodModes.DAYS', function () {
            var momentToUse = moment('2015-01-02'),
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.DAYS, momentToUse)
            expect(isNewGroup).toBeFalsy();
        });

        it('should return true for monday if periodModes.WEEKS', function () {
            var momentToUse = moment('2015-06-01'),//01.06.2015 is a monday
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.WEEKS, momentToUse)
            expect(isNewGroup).toBeTruthy();
        });

        it('should return false for sunday if periodModes.WEEKS', function () {
            var momentToUse = moment('2015-05-31'),
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.WEEKS, momentToUse)
            expect(isNewGroup).toBeFalsy();
        });

        it('should return true for first january if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-01-01'),
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.MONTHS, momentToUse)
            expect(isNewGroup).toBeTruthy();
        });

        it('should return false for thirty first december if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-12-31'),
                isNewGroup;
            isNewGroup = period._isNewPeriodGroup(periodModes.MONTHS, momentToUse)
            expect(isNewGroup).toBeFalsy();
        });

    });

    describe('setPeriodMinimum', function () {
        it('should set monday for provided date if periodModes.WEEKS', function () {
            var momentToUse = moment('2015/02/15');

            period._setPeriodMinimum(momentToUse, periodModes.WEEKS);

            expect(momentToUse.weekday()).toEqual(0);
            expect(momentToUse.date()).toEqual(9);

        });

        it('should set first of month for provided date if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-02-17');

            period._setPeriodMinimum(momentToUse, periodModes.MONTHS);

            expect(momentToUse.date()).toEqual(1);
        });

        it('should set first of january for provided date in first quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-02-17');

            period._setPeriodMinimum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(1);
            expect(momentToUse.month()).toEqual(0);
        });

        it('should set first of april for provided date in second quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-05-17');

            period._setPeriodMinimum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(1);
            expect(momentToUse.month()).toEqual(3);
        });

        it('should set first of januar for provided date if periodModes.YEARS', function () {
            var momentToUse = moment('2015-05-17');

            period._setPeriodMinimum(momentToUse, periodModes.YEARS);

            expect(momentToUse.date()).toEqual(1);
            expect(momentToUse.month()).toEqual(0);
        });
    });

    describe('setPeriodMaximum', function () {
        it('should set sunday for provided date if periodModes.WEEKS', function () {
            var momentToUse = moment('2015-02-17');

            period._setPeriodMaximum(momentToUse, periodModes.WEEKS);

            expect(momentToUse.weekday()).toEqual(6);
            expect(momentToUse.date()).toEqual(22);
        });

        it('should set 31th for provided date in january if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-01-17');

            period._setPeriodMaximum(momentToUse, periodModes.MONTHS);

            expect(momentToUse.date()).toEqual(31);
        });

        it('should set 28th for provided date in february if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-02-17');

            period._setPeriodMaximum(momentToUse, periodModes.MONTHS);

            expect(momentToUse.date()).toEqual(28);
        });

        it('should set 30th for provided date in april if periodModes.MONTHS', function () {
            var momentToUse = moment('2015-04-17');

            period._setPeriodMaximum(momentToUse, periodModes.MONTHS);

            expect(momentToUse.date()).toEqual(30);
        });

        it('should set last of march for provided date in first quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-02-17');

            period._setPeriodMaximum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(31);
            expect(momentToUse.month()).toEqual(2);
        });

        it('should set last of june for provided date in second quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-05-17');

            period._setPeriodMaximum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(30);
            expect(momentToUse.month()).toEqual(5);
        });

        it('should set last of september for provided date in third quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-08-17');

            period._setPeriodMaximum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(30);
            expect(momentToUse.month()).toEqual(8);
        });

        it('should set last of december for provided date in fouth quarter if periodModes.QUARTERS', function () {
            var momentToUse = moment('2015-11-17');

            period._setPeriodMaximum(momentToUse, periodModes.QUARTERS);

            expect(momentToUse.date()).toEqual(31);
            expect(momentToUse.month()).toEqual(11);
        });

        it('should set 31th of december for provided date if periodModes.YEARS', function () {
            var momentToUse = moment('2015-05-17');

            period._setPeriodMaximum(momentToUse, periodModes.YEARS);

            expect(momentToUse.date()).toEqual(31);
            expect(momentToUse.month()).toEqual(11);
        });
    });

    describe('expandRangeToCompletePeriods', function () {
        it('should expand one day to one week if  periodModes.WEEKS', function () {
            var startMoment = moment('2015-05-18');//monday
            var endMoment = moment('2015-05-18');

            period._expandRangeToCompletePeriods(startMoment, endMoment, periodModes.WEEKS);

            //Start momonet should not change
            expect(startMoment.date()).toEqual(18);
            expect(startMoment.month()).toEqual(4);

            //End moment should be end of week
            expect(endMoment.date()).toEqual(24);
            expect(endMoment.month()).toEqual(4);
        });

        it('should expand one day to one month if periodModes.MONTH', function () {
            var startMoment = moment('2015-05-18');//monday
            var endMoment = moment('2015-05-18');

            period._expandRangeToCompletePeriods(startMoment, endMoment, periodModes.MONTHS);

            expect(startMoment.date()).toEqual(1);
            expect(startMoment.month()).toEqual(4);

            expect(endMoment.date()).toEqual(31);
            expect(endMoment.month()).toEqual(4);
        });

        it('should expand one day in second quarter to complete third quarter if periodModes.QUARTERS', function () {
            var startMoment = moment('2015-05-18');//monday
            var endMoment = moment('2015-05-18');

            period._expandRangeToCompletePeriods(startMoment, endMoment, periodModes.QUARTERS);

            expect(startMoment.date()).toEqual(1);
            expect(startMoment.month()).toEqual(3);

            expect(endMoment.date()).toEqual(30);
            expect(endMoment.month()).toEqual(5);
        });
    });

});

describe('period', function () {

    it('should be defined', function () {
        expect(period).toBeDefined();
    });

});

describe('Period', function () {
    var startDate,
        endDate,
        periodInstance;

    //Test data
    var startDateYear = 2015,
        startDateMonth = 1,
        startDateDay = 1,
        endDateYear = 2015,
        endDateMonth = 12,
        endDateDay = 31;

    function instanciatePeriod(periodMode) {
        periodInstance = period.createPeriod(periodMode, startDate, endDate);
    }

    beforeEach(function () {
        startDate = new Date(startDateYear, startDateMonth, startDateDay);
        endDate = new Date(endDateYear, endDateMonth, endDateDay);
    });

    it('createPeriod should return instance of Period', function () {
        instanciatePeriod(periodModes.DAYS);

        expect(periodInstance).toBeDefined();
        expect(periodInstance.start.date()).toEqual(startDateDay);
        expect(periodInstance.end.date()).toEqual(endDateDay);
    });

    it('getValueAsObjects should return correct number of days if PeriodMode is day', function () {
        instanciatePeriod(periodModes.DAYS);

        var days = periodInstance.getValueAsObjects();
        expect(days.length).toEqual(365);
    });

    it('getValueAsObjects should return correct days if PeriodMode is day', function () {
        instanciatePeriod(periodModes.DAYS);

        var days = periodInstance.getValueAsObjects(),
            numberOfFirsts = countDatesInMoments(days, 1),
            numberOfThirtyFirsts = countDatesInMoments(days, 31);

        expect(numberOfFirsts).toEqual(12);
        expect(numberOfThirtyFirsts).toEqual(7);
    });

    it('getValueAsObjects should return correct number of month if PeriodMode is month', function () {
        instanciatePeriod(periodModes.MONTHS);

        var months = periodInstance.getValueAsObjects();
        expect(months.length).toEqual(12);
    });

    it('getValueAsObjects should return correct months if PeriodMode is month', function () {
        instanciatePeriod(periodModes.MONTHS);
        var months = periodInstance.getValueAsObjects();
        for (var i = 1; i < 12; i++) {
            expect(countMonthsInMoments(months, i)).toEqual(1);
        }
    });

    it('isEqual should return true when both periods have the same start, end and periodMode', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate),
            period2 = period.createPeriod(periodModes.DAYS, startDate, endDate);

        expect(period1.isEqual(period2)).toBeTruthy();
    });

    it('isEqual should return false when both periods have the different start', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate),
            falseDate = new Date(2000, 1, 1),
            period2 = period.createPeriod(periodModes.DAYS, falseDate, endDate);

        expect(period1.isEqual(period2)).toBeFalsy();
    });

    it('isEqual should return false when both periods have the different end', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate),
            falseDate = new Date(2000, 1, 1),
            period2 = period.createPeriod(periodModes.DAYS, startDate, falseDate);

        expect(period1.isEqual(period2)).toBeFalsy();
    });

    it('isEqual should return false when both periods have the different periodMode', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate),
            period2 = period.createPeriod(periodModes.MONTHS, startDate, endDate);

        expect(period1.isEqual(period2)).toBeFalsy();
    });

    it('isDirty should be false if date has been created', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate);

        expect(period1.isDirty()).toBeFalsy();
    });

    it('isDirty should be true if date has been altered', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate);

        expect(period1.isDirty()).toBeFalsy();
        period1.start = moment(endDate);
        expect(period1.isDirty()).toBeTruthy();
    });

    it('values should update if date has been altered', function () {
        var period1 = period.createPeriod(periodModes.DAYS, startDate, endDate);
        var valuesBeforeChange = period1.getValueAsObjects(),
            valuesAfterChange;

        period1.start = moment(endDate);
        valuesAfterChange = period1.getValueAsObjects();

        expect(valuesBeforeChange.length).not.toEqual(valuesAfterChange.length);

    });

});


gulpfile.js
'use strict';

var gulp = require('gulp');
var plugins = require('gulp-load-plugins')();
var browserify = require('browserify');
var source = require('vinyl-source-stream');

var paths = {
  lint: ['./gulpfile.js', './lib/**/*.js'],
  watch: ['./gulpfile.js', './lib/**', './test/**/*.js', '!test/{temp,temp/**}'],
  tests: ['./test/**/*.js', '!test/{temp,temp/**}'],
  source: ['./lib/*.js'],
  build: './dist',
  module: './lib/period.js'
};

var filenames = {
  build: {
    scripts: 'period.js'
  },
  release: {
    scripts: 'period.min.js',
    map: './dist/period.map.json'
  }
};

var plumberConf = {};

if (process.env.CI) {
  plumberConf.errorHandler = function (err) {
    throw err;
  };
}

gulp.task('lint', function () {
  return gulp.src(paths.lint)
    .pipe(plugins.jshint('.jshintrc'))
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.jscs())
    .pipe(plugins.jshint.reporter('jshint-stylish'));
});

gulp.task('istanbul', function (cb) {
  gulp.src(paths.source)
    .pipe(plugins.istanbul()) // Covering files
    .pipe(plugins.istanbul.hookRequire()) // Force `require` to return covered files
    .on('finish', function () {
      gulp.src(paths.tests)
        .pipe(plugins.plumber(plumberConf))
        .pipe(plugins.mocha())
        .pipe(plugins.istanbul.writeReports()) // Creating the reports after tests runned
        .on('finish', function () {
          process.chdir(__dirname);
          cb();
        });
    });
});

gulp.task('browserify', function () {
  return browserify({
    debug: true,
    entries: [paths.module]
  })
    .plugin('minifyify', {map: filenames.release.map, output: filenames.release.map})
    .bundle()
    .pipe(source(filenames.release.scripts))
    .pipe(gulp.dest(paths.build));
});

gulp.task('unitTest', function () {
  gulp.src(paths.tests, {cwd: __dirname})
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.jasmine({verbose: true, includeStackTrace: true}));
});

gulp.task('bump', ['test'], function () {
  var bumpType = plugins.util.env.type || 'patch'; // major.minor.patch

  return gulp.src(['./package.json', './bower.json'])
    .pipe(plugins.bump({type: bumpType}))
    .pipe(gulp.dest('./'));
});

gulp.task('watch', ['test'], function () {
  gulp.watch(paths.watch, ['test']);
});

gulp.task('test', ['lint', 'unitTest']);

gulp.task('release', ['bump']);

gulp.task('default', ['test']);

gulp.task('build', ['browserify']);


lib/period.js
/*
 * gfk-period
 *
 * Copyright (c) 2015
 */
'use strict';
/*global module, define, require */
(function (root, factory) {
    if (typeof module !== 'undefined' && module.exports) {
        // CommonJS
        module.exports = factory(require('moment'), require('moment-range'));
    } else if (typeof define === 'function' && define.amd) {
        // AMD
        define(['moment'], factory);
    } else {
        // Global Variables
        root.period = factory(root.moment);
    }
}(this, function (moment) {
    moment.locale('de', {
        week: {
            dow: 1, // Monday is the first day of the week.
            doy: 4  // The week that contains Jan 4th is the first week of the year.
        }
    });

    var api,
        Period,
        PERIOD_MODES = {
            DAYS: 'd',
            WEEKS: 'w',
            MONTHS: 'm',
            QUARTERS: 'q',
            YEARS: 'y'
        },
        checksumFormat='YYYYMMDD';

    /**
     * A Period describes a time range from end to start and provides all dates
     * in the period in different formats
     * The provided dates are always expanded to stretch over complete periods
     * @param {string} periodMode Describes which time unit the period uses
     * @param {date|moment} start The date on which the period starts
     * @param {date|moment} end The date on which the period ends
     * @constructor
     */
    Period = function (periodMode, start, end) {
        //TODO:Make variables private and expose setters and getters
        this.periodMode = periodMode;
        this.start = moment.utc(start);
        this.end = moment.utc(end);

        expandRangeToCompletePeriods(this.start, this.end, this.periodMode);

        this.values = null;
        this.checksum = this.getChecksum();

    };

    Period.prototype.getLongPeriodLabel = function () {
        var labelStart = this.getLongStringForStart(),
            labelEnd = this.getLongStringForEnd();

        return labelStart + ' - ' + labelEnd;
    };

    Period.prototype.isEqual = function (periodToCompare) {
        var hasSameStart,
            hasSameEnd,
            hasSamePeriodMode;

        hasSameStart = this.start.isSame(periodToCompare.start);
        hasSameEnd = this.end.isSame(periodToCompare.end);
        hasSamePeriodMode = this.periodMode === periodToCompare.periodMode;

        return hasSameStart && hasSameEnd && hasSamePeriodMode;
    };

    Period.prototype.getChecksum = function () {
        var startCheckSum = this.start.format(checksumFormat),
            endCheckSum = this.end.format(checksumFormat);

        return this.periodMode + '/' + startCheckSum + '/' + endCheckSum;
    };

    Period.prototype.isDirty = function () {
        return this.checksum !== this.getChecksum();
    };

    Period.prototype.getLongStringForStart = function () {
        return this.start.format(getLongPeriodFormat(this.periodMode));
    };

    Period.prototype.getLongStringForEnd = function () {
        return this.end.format(getLongPeriodFormat(this.periodMode));
    };

    Period.prototype.getLongStringFormat = function () {
        return getLongPeriodFormat(this.periodMode);
    };

    Period.prototype.getShortStringFormat = function () {
        return getShortPeriodFormat(this.periodMode);
    };

    Period.prototype.getGroupStringFormat = function () {
        return getPeriodGroupFormat(this.periodMode);
    };

    /**
     * Returns array of DTOs with all dates the period consists of, grouped by its PeriodMode
     * Checks if the start and end values have been altered and recalculates if so
     * @returns array {{key: (string), value: (moment)}}
     */
    Period.prototype.getValueAsObjects = function () {
        if (this.values !== null && this.isDirty() === false) {
            return this.values;
        }
        var periodInstance = this,
            momentRangeIterator = getMomentRangeIterator(this.periodMode),
            currentRange = moment().range(this.start, this.end);


        this.values = [];
        currentRange.by(momentRangeIterator, function (momentToUse) {
            momentToUse.isNewGroup = isNewPeriodGroup(periodInstance.periodMode, momentToUse);
            momentToUse.checksum = momentToUse.format(checksumFormat);
            periodInstance.values.push(momentToUse);
        });

        this.checksum = this.getChecksum();
        return this.values;
    };

    function expandRangeToCompletePeriods(start, end, periodMode) {
        setPeriodToMinimum(start, periodMode);
        setPeriodToMaximum(end, periodMode);
    }

    function setPeriodToMinimum(momentToUse, periodMode) {
        switch (periodMode) {
            case PERIOD_MODES.WEEKS:
                momentToUse.weekday(0);
                break;
            case PERIOD_MODES.MONTHS:
                momentToUse.date(1);
                break;
            case PERIOD_MODES.QUARTERS:
                momentToUse.date(1);
                switch (momentToUse.quarter()) {
                    case 1:
                        momentToUse.month(0);
                        break;
                    case 2:
                        momentToUse.month(3);
                        break;
                    case 3:
                        momentToUse.month(6);
                        break;
                    case 4:
                        momentToUse.month(9);
                        break;
                }
                break;
            case PERIOD_MODES.YEARS:
                momentToUse.date(1);
                momentToUse.month(0);
                break;
        }
    }

    function setPeriodToMaximum(momentToUse, periodMode) {
        switch (periodMode) {
            case PERIOD_MODES.WEEKS:
                momentToUse.weekday(6);
                break;
            case PERIOD_MODES.MONTHS:
                momentToUse.date(1).add(1, 'months').subtract(1, 'days');
                break;
            case PERIOD_MODES.QUARTERS:
                //When setting dates to maximum, always set the month first
                // else you might set it to a higher date than the current month
                // and it will bubble up and mess up the date
                switch (momentToUse.quarter()) {
                    case 1:
                        momentToUse.month(2);
                        momentToUse.date(31);
                        break;
                    case 2:
                        momentToUse.month(5);
                        momentToUse.date(30);
                        break;
                    case 3:
                        momentToUse.month(8);
                        momentToUse.date(30);
                        break;
                    case 4:
                        momentToUse.month(11);
                        momentToUse.date(31);
                        break;
                }
                break;
            case PERIOD_MODES.YEARS:
                momentToUse.month(11);
                momentToUse.date(31);
                break;
        }
    }

    /**
     * Create a new key/value object from give moment with provided period set to a new value
     * Used as DTO for the front end
     * @param {string} periodMode Period which will be altered
     * @param {moment} momentToUse Moment object that will be cloned and formatted (provided moment will not be altered)
     * @param {number} [newValue=null] number the given period will be set to
     * @returns {{key: (string), value: (moment)}}
     */
    function createDateObject(periodMode, momentToUse, newValue) {
        var formatToUse = getShortPeriodFormat(periodMode),
            clonedMoment = moment(momentToUse);

        if (newValue) {
            setPeriodOnMoment(periodMode, clonedMoment, newValue);
        }

        return {
            key: clonedMoment.format(formatToUse),
            value: clonedMoment
        };
    }

    function getShortPeriodFormat(periodMode) {
        if (periodMode === 'ytd') {
            return '[YTD] YY';
        }
        switch (periodMode[0]) {
            case PERIOD_MODES.DAYS:
                return 'DD.';
            case PERIOD_MODES.WEEKS:
                return '[KW] W GGGG';
            case PERIOD_MODES.MONTHS:
                return 'MMMM YYYY';
            case PERIOD_MODES.QUARTERS:
                return '[Q.]Q YYYY';
            case PERIOD_MODES.YEARS:
                return 'YYYY';
            case PERIOD_MODES.TOTAL:
                return '[Total]';
        }
        throw 'No short period format found for "' + periodMode + '"';
    }

    function getLongPeriodFormat(periodMode) {
        if (periodMode === 'ytd') {
            return '[YTD] YYYY';
        }
        switch (periodMode[0]) {
            case PERIOD_MODES.DAYS:
                return 'DD.MM.YYYY';
            case PERIOD_MODES.WEEKS:
                return '[KW] W GGGG';
            case PERIOD_MODES.MONTHS:
                return 'MMMM YYYY';
            case PERIOD_MODES.QUARTERS:
                return '[Q.]Q YYYY';
            case PERIOD_MODES.YEARS:
                return 'YYYY';
            case PERIOD_MODES.TOTAL:
                return '[Total]';
        }
        throw 'No long period format found for "' + periodMode + '"';
    }

    function getPeriodGroupFormat(periodMode) {
        if (periodMode === 'ytd') {
            return '[YTD] YYYY';
        }
        switch (periodMode[0]) {
            case PERIOD_MODES.DAYS:
                return 'DD. MMMM YYYY';
            case PERIOD_MODES.WEEKS:
                return '[KW] W GGGG';
            case PERIOD_MODES.MONTHS:
                return 'MMMM YYYY';
            case PERIOD_MODES.QUARTERS:
                return '[Quartal] Q YYYY';
            case PERIOD_MODES.YEARS:
                return 'YYYY';
            case PERIOD_MODES.TOTAL:
                return '[Total]';
        }
        throw 'No long period format found for "' + periodMode + '"';
    }

    function getMomentSetterFunction(periodMode) {
        switch (periodMode) {
            case PERIOD_MODES.DAYS:
                return 'date';
            case PERIOD_MODES.WEEKS:
                return 'week';
            case PERIOD_MODES.MONTHS:
                return 'month';
            case PERIOD_MODES.QUARTERS:
                return 'quarter';
            case PERIOD_MODES.YEARS:
                return 'year';
        }
        throw 'No moment setter found for period "' + periodMode + '"';
    }

    /**
     * Show if the group of the date change according to given periodMode
     * Days are grouped in months, so this will return true on the first of every month
     * @param {string} periodMode
     * @param {moment} momentToUse
     * @returns {boolean}
     */
    function isNewPeriodGroup(periodMode, momentToUse) {
        switch (periodMode) {
            case PERIOD_MODES.DAYS:
                return momentToUse.date() === 1;
            case PERIOD_MODES.WEEKS:
                return momentToUse.week() === 1;
            case PERIOD_MODES.MONTHS:
            case PERIOD_MODES.QUARTERS:
                return momentToUse.dayOfYear() === 1;
            case PERIOD_MODES.YEARS:
                return true;
        }
        throw 'No newGroup rule found for period "' + periodMode + '"';
    }

    function getMomentRangeIterator(periodMode) {
        switch (periodMode) {
            case PERIOD_MODES.DAYS:
                return 'days';
            case PERIOD_MODES.WEEKS:
                return 'weeks';
            case PERIOD_MODES.MONTHS:
                return 'months';
            case PERIOD_MODES.QUARTERS:
                return 'quarters';
            case PERIOD_MODES.YEARS:
                return 'years';
        }
        throw 'No moment setter found for period "' + periodMode + '"';
    }

    function setPeriodOnMoment(periodMode, moment, value) {
        var setterFunction = getMomentSetterFunction(periodMode);
        moment[setterFunction](value);
    }

    /**
     * Get the count of days in a specific month
     * @param {number} year Year of the date to be used
     * @param {number} month Zero based month to be used
     * @returns {number}
     */
    function getNumberOfDaysInMonth(year, month) {
        return new Date(year, month + 1, 0).getDate();
    }

    api = {};

    /**
     * Create a new instance of Period
     * @param {string} periodMode PeriodMode of this instance
     * @param {date|moment} start The date on which the period starts
     * @param {date|moment} end The date on which the period ends
     * @returns {Period}
     */
    api.createPeriod = function (periodMode, start, end) {
        return new Period(periodMode, start, end);
    };

    api.PERIOD_MODES = PERIOD_MODES;

    api.getMomentRangeIterator = getMomentRangeIterator;

//Expose privates for testing
    api._getNumberOfDaysInMonth = getNumberOfDaysInMonth;
    api._setPeriodOnMoment = setPeriodOnMoment;
    api._isNewPeriodGroup = isNewPeriodGroup;
    api._setPeriodMinimum = setPeriodToMinimum;
    api._setPeriodMaximum = setPeriodToMaximum;
    api._expandRangeToCompletePeriods = expandRangeToCompletePeriods;

    return api;

}));
